Validate disk offering IOPS normal and maximum read/write values

GabrielBrascher opened the pull request:
## Description

When creating a new disk offering ('createDiskOffering' command on API) one can set _normal_ IOPS (write/read) higher than the maximum, breaking the logics on the disk offerings. Additionally, one can (by mistake) put irrealistic values for the bursting length. Example if adding few zeros by mistake and configuring **360000** seconds, which turns out to be a bursting length of **100** hours.

This PR adds verifications ensuring that either Bytes or IOPS bursts cannot be greater than a configured value of the following global settings variables.

```
vm.disk.bytes.maximum.read.length
vm.disk.bytes.maximum.write.length
vm.disk.iops.maximum.read.length
vm.disk.iops.maximum.write.length
```

The default value is 0 (zero), allows any IOPS maximum rate length; thus keeping backward compatibility and not affecting any environment configuration after upgrades.

# Example:
One can set the 'vm.disk.bytes.maximum.read.length' to '3600' (seconds); once the parameter is updated on the management server it will not be possible to create disk offerings with IOPS length higher than 3600 seconds (60 minutes).

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)g

GabrielBrascher commented:
@DaanHoogland I totally agree with you, that is why 'iops.maximum.rate.length' is zero by default (keeping the current behavior).  The idea is that system admins should also be able to enable/disable some restrictions, for safety.

The reason for this PR is because I have seen a case where normal IOPS was configured (by mistake) higher than Maximum IOPS. System admins should have control and be able to execute any action; however, humans are always error-prone :slightly_smiling_face:

DaanHoogland commented:
makes sense @GabrielBrascher the check for those values should be enough, though. setting the extra maximum rate length seems excessive.

wido commented:
@DaanHoogland The problem is that libvirt/qemu might trip over a too high value causing Instances not be started with a very odd Exception.

I'd personally would set the maximum burst to 3600 (1h) to make sure typos won't prevent a VM from starting.

The same goes with sanity checking. It's Libvirt/Qemu which refuses to start the VM if these values are wrong.

DaanHoogland commented:
@blueoranutang package

GabrielBrascher commented:
@DaanHoogland small typo on the blueorangutan package call. Could you please recall 
@blueorangutan package? :)

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

GabrielBrascher commented:
Looks like _blueoragutan_ listens to me as well for packaging, I am not able to call for tests then. Thanks anyway @DaanHoogland.

blueorangutan commented:
Packaging result: ✖centos6 ✔centos7 ✔debian. JID-395

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✔centos7 ✔debian. JID-400

andrijapanicsb commented:
@GabrielBrascher - just FMI - when/where was this burst length added? I mean in which Qemu version?
That's part of blkdeviotune in qemu/libvirt,  I suppose?

GabrielBrascher commented:
@andrijapanicsb Qemu supports IOPS bursts since version 2.6:
- https://blogs.igalia.com/berto/2016/05/24/io-bursts-with-qemu-2-6/
- https://github.com/qemu/qemu/blob/stable-2.6/docs/throttle.txt
- https://wiki.qemu.org/ChangeLog/2.6

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-566)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 39627 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3681-t566-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Smoke tests completed. 77 look OK, 0 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

GabrielBrascher commented:
Thanks for the review and tests, @DaanHoogland!

andrijapanicsb commented:
@GabrielBrascher this looks fine. However, I would ask if it makes sense (or not!) to also validate not only the iops burst duration but also bytes rate duration? If not, that's all OK - let me know, please.

i.e.

> (localcloud) SBCM5> > list configurations name=iops.maximum.rate.length | grep value\"
      "value": "10"
(localcloud) SBCM5> > create diskoffering bytesreadrate=1000 bytesreadratemax=1500 byteswriterate=1000 byteswriteratemax=1500 iopsreadrate=2000 iopsreadratemax=2500 iopswriterate=2000 iopswriteratemax=2500 bytesreadratemaxlength=15 byteswriteratemaxlength=15 iopsreadratemaxlength=15 iopswriteratemaxlength=15 name=xxx displaytext=xxx disksize=1
 Error: (HTTP 431, error code 4350) IOPS read max length (15 seconds) cannot be greater than iops.maximum.rate.length (10s seconds)
(localcloud) SBCM5> > create diskoffering bytesreadrate=1000 bytesreadratemax=1500 byteswriterate=1000 byteswriteratemax=1500 iopsreadrate=2000 iopsreadratemax=2500 iopswriterate=2000 iopswriteratemax=2500 bytesreadratemaxlength=15 byteswriteratemaxlength=15 iopsreadratemaxlength=15 iopswriteratemaxlength=9 name=xxx displaytext=xxx disksize=1
 Error: (HTTP 431, error code 4350) IOPS read max length (15 seconds) cannot be greater than iops.maximum.rate.length (10s seconds)
(localcloud) SBCM5> > create diskoffering bytesreadrate=1000 bytesreadratemax=1500 byteswriterate=1000 byteswriteratemax=1500 iopsreadrate=2000 iopsreadratemax=2500 iopswriterate=2000 iopswriteratemax=2500 bytesreadratemaxlength=15 byteswriteratemaxlength=15 iopsreadratemaxlength=9 iopswriteratemaxlength=9 name=xxx displaytext=xxx disksize=1
{
  "diskoffering": {
    "created": "2020-01-13T21:37:38+0000",
    "diskBytesReadRate": 1000,
    "diskBytesReadRateMax": 1500,
    "diskBytesReadRateMaxLength": 15,
    "diskBytesWriteRate": 1000,
    "diskBytesWriteRateMax": 1500,
    "diskBytesWriteRateMaxLength": 15,
    "diskIopsReadRate": 2000,
    "diskIopsReadRateMax": 2500,
    "diskIopsReadRateMaxLength": 9,
    "diskIopsWriteRate": 2000,
    "diskIopsWriteRateMax": 2500,
    "diskIopsWriteRateMaxLength": 9,
    "disksize": 1,
    "displayoffering": true,
    "displaytext": "xxx",
    "id": "3920175d-ee3e-46d4-b25c-3d12fb040d25",
    "iscustomized": false,
    "name": "xxx",
    "provisioningtype": "thin",
    "storagetype": "shared"
  }
}

wido commented on server/src/test/java/com/cloud/configuration/ConfigurationManagerTest.java:
```diff
@@ -166,6 +168,8 @@
 
     VlanVO vlan = new VlanVO(Vlan.VlanType.VirtualNetwork, "vlantag", "vlangateway", "vlannetmask", 1L, "iprange", 1L, 1L, null, null, null);
 
+    private static final String MAXIMUM_IOPS_ALLOWED = "3600";
```
Maybe rename this to MAXIMUM_LENGTH_ALLOWED as it's a time which we specify here?

DaanHoogland replied:
imho: MAXIMUM_DURATION_ALLOWED

wido commented on server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:
```diff
@@ -2975,6 +2983,49 @@ public DiskOffering createDiskOffering(final CreateDiskOfferingCmd cmd) {
+            }
+        }
+    }
+
```
As suggested I would also validate the bytes duration here

wido commented on server/src/test/java/com/cloud/configuration/ConfigurationManagerTest.java:
```diff
@@ -854,36 +848,43 @@ public void hasSameSubnetTest() {
+                    "2001:db9:0:f101::2", ipV6Network);
             Assert.fail();
-        } catch(InvalidParameterValueException e){
+        } catch (InvalidParameterValueException e) {
```
See all the lines above. Maybe revert these checkstyle changes?

GabrielBrascher commented:
Thanks for the feedback @andrijapanicsb @wido! Code has been updated.

weizhouapache commented:
@GabrielBrascher thanks for your work.

There are some global settings related to vm disk throttling, it would be better to take them into consideration.
https://github.com/apache/cloudstack/blob/master/server/src/main/java/com/cloud/configuration/Config.java#L834-L865

svenvogel commented:
@GabrielBrascher Hi Gabriel. Nice feature! we need to add this to primate too ... i will take care ... 😀

GabrielBrascher commented:
Cool @svenvogel! Thanks :-)

GabrielBrascher commented on server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:
```diff
@@ -389,6 +389,18 @@
+                    + " Example: iops.maximum.rate.length = 3600 sets the maximum IOPS length accepted for a disk offering as 3600 seconds (60 minutes).",
+            false, ConfigKey.Scope.Global, null);
+
+    public static ConfigKey<Long> iopsMaximumBytes = new ConfigKey<Long>(Long.class, "iops.maximum.rate.length", "Advanced", "0",
```
I don't see any problem on having maximum iops for write and read separately if this makes sense on a practical way. What do you think @wido?

GabrielBrascher commented on server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:
```diff
@@ -6252,6 +6325,6 @@ public String getConfigComponentName() {
     @Override
     public ConfigKey<?>[] getConfigKeys() {
-        return new ConfigKey<?>[] {SystemVMUseLocalStorage};
+        return new ConfigKey<?>[] {SystemVMUseLocalStorage, iopsMaximumRateLength};
```
You are right, @weizhouapachem. That slipped on my sight after adding the new config key when addressing @andrijapanicsb review. Thanks!

DaanHoogland commented:
rekick jenkins

GabrielBrascher commented on server/src/main/java/com/cloud/configuration/Config.java:
```diff
@@ -861,31 +861,31 @@
     VmDiskThrottlingIopsReadRate(
             "Advanced",
             ManagementServer.class,
-            Integer.class,
```
Not sure why it was set as Integer. All the pieces of code using these configs are parsing them as Long.

GabrielBrascher commented:
@weizhouapache thanks for the review, I pushed a commit tackling all of them.

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✖centos7 ✖debian. JID-703

GabrielBrascher commented:
@rhtyd found some checkstyle issues on this PR, my dev environment had a few misconfigurations and caused them. I will fix it in a couple of minutes.

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✔centos7 ✔debian. JID-708

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-849)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 30902 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3681-t849-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_disk_offerings.py
Smoke tests completed. 77 look OK, 1 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_05_create_burst_type_disk_offering | `Error` | 0.04 | test_disk_offerings.py

DaanHoogland commented:
all of a sudden, some test fail that look like they might be related: test_05_create_burst_type_disk_offering can you double check the logs and run those locally, @GabrielBrascher ?

GabrielBrascher commented:
@DaanHoogland I am going to check it, thanks for the heads up!

GabrielBrascher commented:
I am going to invest some time today to dig into test_05_create_burst_type_disk_offering. Hope to have an update later today here @DaanHoogland :- )

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✔centos7 ✔debian. JID-756

andrijapanicsb commented:
@blueorangutan test

blueorangutan commented:
@andrijapanicsb a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-926)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 31701 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3681-t926-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_vpc_vpn.py
Smoke tests completed. 79 look OK, 0 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

GabrielBrascher commented:
@andrijapanicsb thanks for running the tests :)
@DaanHoogland it seems that `test_05_create_burst_type_disk_offering` has passed now after commit ae57e4d2e382aaf5c05c716a153b9cbcb2903790.

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✖centos7 ✖debian. JID-814

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✖centos7 ✖debian. JID-858

andrijapanicsb commented:
try tomorrow please, we are handling some stuff internally...

DaanHoogland commented:
@blueorangutan package

blueorangutan commented:
@DaanHoogland a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✖centos7 ✖debian. JID-874

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✖centos7 ✖debian. JID-898

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✔centos7 ✔debian. JID-922

GabrielBrascher commented:
@weizhouapache @DaanHoogland are you ok with this PR?

DaanHoogland commented:
@blueorangutan test matrix

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins matrix job (centos7 mgmt + xs71, centos7 mgmt + vmware65, centos7 mgmt + kvmcentos7) has been kicked to run smoke tests

DaanHoogland commented:
@blueorangutan test centos7 xenserver-71

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + xenserver-71) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-1074)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 30195 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3681-t1074-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_loadbalance.py
Intermittent failure detected: /marvin/tests/smoke/test_privategw_acl.py
Smoke tests completed. 79 look OK, 1 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_02_vpc_privategw_static_routes | `Failure` | 186.52 | test_privategw_acl.py
test_03_vpc_privategw_restart_vpc_cleanup | `Failure` | 181.45 | test_privategw_acl.py
test_04_rvpc_privategw_static_routes | `Failure` | 260.04 | test_privategw_acl.py

blueorangutan commented:
<b>Trillian test result (tid-1075)</b>
Environment: vmware-65u2 (x2), Advanced Networking with Mgmt server 7
Total time taken: 34465 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3681-t1075-vmware-65u2.zip
Intermittent failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermittent failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Smoke tests completed. 74 look OK, 1 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_02_vpc_privategw_static_routes | `Failure` | 341.35 | test_privategw_acl.py
test_03_vpc_privategw_restart_vpc_cleanup | `Failure` | 334.95 | test_privategw_acl.py
test_04_rvpc_privategw_static_routes | `Failure` | 532.00 | test_privategw_acl.py

blueorangutan commented:
<b>Trillian test result (tid-1076)</b>
Environment: xenserver-71 (x2), Advanced Networking with Mgmt server 7
Total time taken: 31420 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3681-t1076-xenserver-71.zip
Intermittent failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermittent failure detected: /marvin/tests/smoke/test_scale_vm.py
Smoke tests completed. 78 look OK, 2 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_02_vpc_privategw_static_routes | `Failure` | 282.78 | test_privategw_acl.py
test_03_vpc_privategw_restart_vpc_cleanup | `Failure` | 237.73 | test_privategw_acl.py
test_04_rvpc_privategw_static_routes | `Failure` | 366.13 | test_privategw_acl.py
test_01_scale_vm | `Failure` | 26.81 | test_scale_vm.py

GabrielBrascher commented:
Ping for review @rhtyd @svenvogel @nvazquez :slightly_smiling_face:
@weizhouapache @andrijapanicsb are you ok with the PR after addressing reviews?

DaanHoogland commented:
@weizhouapache @wido I am fine with this and need a second ok

GabrielBrascher commented:
@DaanHoogland I am doing a small refactoring in this PR. I will update you soon (today). thanks for the reply.

GabrielBrascher commented:
@DaanHoogland @andrijapanicsb @weizhouapache @wido @svenvogel I have updated this PR. Thanks for all the feedback so far.

GabrielBrascher commented:
**### Code Update ###**

After digging around the configurations` vm.disk.throttling.*`, I found out that they were not the desired ones. Therefore, it was necessary to refactor the code. First I reverted a few commits and then implemented the validation with a new set of config keys.

The main goal of this PR is to check the length (seconds) of burst IOPS/Bytes. The `vm.disk.throttling.*` configs consider the rate (e.g. bytes read per second), not length (burst duration)

Before this refactoring the following parameters where being considered:
```
vm.disk.throttling.iops_read_rate
vm.disk.throttling.iops_write_rate
vm.disk.throttling.bytes_read_rate
vm.disk.throttling.bytes_write_rate
```

Current commit considers the following configurations:
```
vm.disk.bytes.maximum.read.length
vm.disk.bytes.maximum.write.length
vm.disk.iops.maximum.read.length
vm.disk.iops.maximum.write.length
```

Additionally, I removed a few checks that did not make much sense; also, I converted a bunch of **if** statements into a small and concise method that is called 4 times, making it easier to test and extend if necessary.

**### Manual tests ###**
Configuration when beginning the tests

	vm.disk.bytes.maximum.read.length 	= 0
	vm.disk.bytes.maximum.write.length 	= 0
	vm.disk.iops.maximum.read.length 	= 0
	vm.disk.iops.maximum.write.length 	= 0

	iops_read_rate = 10000
	iops_read_rate_max = 20000
	iops_read_rate_max_length = 36000

	iops_write_rate = 10000
	iops_write_rate_max = 20000
	iops_write_rate_max_length = 36000

	bytes_read_rate = 10000
	bytes_read_rate_max = 20000
	bytes_read_rate_max_length = 36000

	bytes_write_rate = 10000
	bytes_write_rate_max = 20000
	bytes_write_rate_max_length = 36000	


- **Test 001**
**Input (tests via API with createDiskOffering command):**
	iops_read_rate = 10000
	iops_read_rate_max = 20000
	iops_read_rate_max_length = 36000

	iops_write_rate = 10000
	iops_write_rate_max = 20000
	iops_write_rate_max_length = 36000

	bytes_read_rate = 10000
	bytes_read_rate_max = 20000
	bytes_read_rate_max_length = 36000

	bytes_write_rate = 10000
	bytes_write_rate_max = 20000
	bytes_write_rate_max_length = 36000

**Expected result:** create disk offering
**Result:** All good

- **Test 002**
Input (via API):
	Same as Test 001, except by:
	bytes_write_rate_max_length = -1

**Expected result:** fail during disk offering creation due to bytes_write_rate_max_length = -1
**Result: Error:** (HTTP 431, error code 4350) Bytes Write max length (-1) must be a positive value

- **Test 003**
	Same as Test 001, except by:
	iops_read_rate_max = 100

**Expected result:** fail during disk offering creation due to iops_read_rate (36000) > iops_read_rate_max (100)
**Result:** (HTTP 431, error code 4350) "IOPS Read rate (36000) cannot be greater than IOPS Read maximum rate (100)"

- **Test 003**
	Same as Test 001, except by:
	vm.disk.bytes.maximum.read.length = 123
	(dinamycally changed, does not require restart)

**Expected result:** fail due to bytes_read_rate_max_length (36000) > vm.disk.bytes.maximum.read.length = 123
**Result:** (HTTP 431, error code 4350) "Bytes read max length (36000 seconds) cannot be greater than vm.disk.bytes.maximum.read.length (123 seconds)"


- Test 004
	Same as Test 001, except by:
	vm.disk.bytes.maximum.read.length = 99999999

**Expected result:** create disk offering
**Result:** All good

- **Test 005**
	Same as Test 001, except by:
	vm.disk.iops.maximum.read.length = 39999

**Expected result:** fail due to bytes_read_rate_max_length (36000) > vm.disk.iops.maximum.read.length = 39999
**Result:** (HTTP 431, error code 4350) "IOPS read max length (36000 seconds) cannot be greater than vm.disk.iops.maximum.read.length (35999 seconds)"

Did a few more tests alternating the vm.disk.(*).length parameters

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✔centos7 ✔debian. JID-964

RodrigoDLopez commented:
## MANUAL TEST ##
```
On global settings:
vm.disk.bytes.maximum.read.length = 0
vm.disk.bytes.maximum.write.length = 0
vm.disk.iops.maximum.read.length = 0
vm.disk.iops.maximum.write.length = 0
```

Testing via API
* **test 01**
	bytesreadrate=1000 
	bytesreadratemax=1000 
	bytesreadratemaxlength=2000 

	byteswriterate=1000 
	byteswriteratemax=1000 
	byteswriteratemaxlength=2000 

	iopsreadrate=1000 
	iopsreadratemax=1000 
	iopsreadratemaxlength=2000 

	iopswriterate=1000 
	iopswriteratemax=1000 
	iopswriteratemaxlength=2000

**Expected result:** create the offering
**Result:** All good

* **test 02**
	Same as test 01, except by:
	bytesreadratemax=900

**Expected result:** fail during disk offering creation due to bytes_read_rate (1000) > bytes_read_rate_max (900)
**Result:**  Bytes Read rate (1000) cannot be greater than Bytes Read maximum rate (900)

* **test 03**
	Same as test 01, except by:
	byteswriteratemax=900
	
**Expected result:** fail during disk offering creation due to bytes_write_rate (1000) > bytes_write_rate_max (900)
**Result:**  Bytes Write rate (1000) cannot be greater than Bytes Write maximum rate (900)

* **test 04**
	Same as test 01, except by:
	iopsreadratemax=900
	
**Expected result:** fail during disk offering creation due to iops_read_rate (1000) > iops_read_rate_max (900)
**Result:**  IOPS Read rate (1000) cannot be greater than IOPS Read maximum rate (900)

* **test 05**
	Same as test 01, except by:
	iopswriteratemax=900
	
**Expected result:** fail during disk offering creation due to iops_write_rate (1000) > iops_write_rate_max (900)
**Result:**  IOPS Write rate (1000) cannot be greater than IOPS Write maximum rate (900)

```
On global settings:
vm.disk.bytes.maximum.read.length = 2000
vm.disk.bytes.maximum.write.length = 2000
vm.disk.iops.maximum.read.length = 2000
vm.disk.iops.maximum.write.length = 2000
```

* **test 06**
	same as test 01, except by:
	bytesreadratemaxlength=2001

**Expected result:** fail during disk offering creation due to bytes_read_rate_max_length (2001) >  vm.disk.bytes.maximum.read.length (2000)
**Result:** Bytes read max length (2001 seconds) cannot be greater than vm.disk.bytes.maximum.read.length (2000 seconds)

* **test 07**
	same as test 01, except by:
	byteswriteratemaxlength=2001

**Expected result:** fail during disk offering creation due to bytes_write_rate_max_length (2001) >  vm.disk.bytes.maximum.write.length (2000)
**Result:** All good

* **test 08**
	same as test 01, except by:
	iopsreadratemaxlegth=2001

**Expected result:** fail during disk offering creation due to iops_read_rate_max_length (2001) >  vm.disk.iops.maximum.read.length (2000)
**Result:** IOPS read max length (2001 seconds) cannot be greater than vm.disk.iops.maximum.read.length (2000 seconds)

* **test 09**
	same as test 01, except by:
	iopswriteratemaxlegth=2001

**Expected result:** fail during disk offering creation due to iops_write_rate_max_length (2001) >  vm.disk.iops.maximum.write.length (2000)
**Result:** IOPS write max length (2001 seconds) cannot be greater than vm.disk.iops.maximum.write.length (2000 seconds)

The 'test 07' result is all good, except by: it should have failed
I requested some changes to fix this.

GabrielBrascher commented on server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:
```diff
@@ -2999,6 +3020,57 @@ public DiskOffering createDiskOffering(final CreateDiskOfferingCmd cmd) {
+        }
+
+        if (BYTES_MAX_WRITE_LENGTH.value() != null && BYTES_MAX_WRITE_LENGTH.value() != 0l) {
+            if (bytesReadRateMaxLength != null && bytesReadRateMaxLength > BYTES_MAX_WRITE_LENGTH.value()) {
```
Well noted, thanks for the review and tests, @RodrigoDLopez. I just fixed it!

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

svenvogel commented:
@GabrielBrascher from logic it looks good i think.

```log
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project cloud-engine-schema: Compilation failure -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
```

what that for an jenkins error?

GabrielBrascher commented:
@svenvogel I checked the Jenkins stack and there are not many details. There is a compilation failure at project cloud-engine-schema. I will rebase this branch with the current master. Thanks for the ping.

The good news is that I could build it without any problem. @RodrigoDLopez also was able to build and test it.

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✖centos7 ✔debian. JID-976

blueorangutan commented:
Packaging result: ✖centos6 ✖centos7 ✔debian. JID-977

GabrielBrascher commented:
@blueorangutan package

blueorangutan commented:
@GabrielBrascher a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✔centos7 ✔debian. JID-979

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-1168)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 29506 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3681-t1168-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_diagnostics.py
Smoke tests completed. 80 look OK, 0 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

DaanHoogland commented:
@wido @weizhouapache @andrijapanicsb can we merge?

rhtyd commented on server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:
```diff
@@ -400,6 +400,20 @@
     public static final ConfigKey<Boolean> SystemVMUseLocalStorage = new ConfigKey<Boolean>(Boolean.class, "system.vm.use.local.storage", "Advanced", "false",
             "Indicates whether to use local storage pools or shared storage pools for system VMs.", false, ConfigKey.Scope.Zone, null);
 
+    public final static ConfigKey<Long> BYTES_MAX_READ_LENGTH= new ConfigKey<Long>(Long.class, "vm.disk.bytes.maximum.read.length", "Advanced", "0",
```
I'm not sure how to test/review this. Should this be checked by the storage driver or hypervisor plugin?

GabrielBrascher replied:
@rhtyd sorry but I'm not sure if I understood your doubt. Do you think that those parameters could be placed in another class?

This implementation covers a few validations before creating the disk offering; therefore I placed the validation code and parameters at this class, right before the execution of `com.cloud.configuration.ConfigurationManagerImpl.createDiskOffering(...)` [1].

One of the validation avoids unrealistic burst length (e.g. hours of writing/reading burst). On the first proposal, I simply added a final static variable avoiding one hour of burst; however decided later with the help of reviewers to add parameters and leave it as configurable. Default values do not change the current behavior.

https://github.com/apache/cloudstack/blob/debe056beeff3aea67ab5c756158a60c25bced96/server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java#L3021

blueorangutan commented:
<b>Trillian test result (tid-1222)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 28800 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr3681-t1222-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_hostha_kvm.py
Smoke tests completed. 80 look OK, 0 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

GabrielBrascher commented:
Hi @borisstoyanov , thanks for testing this PR. The global settings affect the following disk offering parameters:

```
bytesreadratemaxlength
byteswriteratemaxlength
iopsreadratemaxlength
iopswriteratemaxlength

```
The setup on global settings are according to the following parameters:
```
vm.disk.bytes.maximum.read.length = 0
vm.disk.bytes.maximum.write.length = 0
vm.disk.iops.maximum.read.length = 0
vm.disk.iops.maximum.write.length = 0

```

Quick resume on the length parameters. The verifications ensure that either Bytes or IOPS bursts cannot be greater than a configured value of those global settings variables.

Additionally, this PR checks the normal IOPS (write/read) ensuring that it cannot be higher than the respective maximum value (if the maximum has been set by the operator, otherwise, if null, it ignores the maximum).

Example of all parameters necessary to check if length (e.g. bytesreadratemaxlength) and normal (e.g. bytesreadrate) vs maximum (bytesreadratemax) are OK:

- Bytes (read / right)
```
bytesreadrate, bytesreadratemax, bytesreadratemaxlength
byteswriterate, byteswriteratemax, byteswriteratemaxlength
```

- IOPS (read / right)
```
iopsreadrate, iopsreadratemax, iopsreadratemaxlength
iopswriterate, iopswriteratemax, iopswriteratemaxlength
```

Therefore, the test that you presented should indeed accept the deploy as you have done (no length neither maximum IOPS/bytes have been configured via the disk offering command).

Once again, thanks for checking this PR, in case of any question just ping me.

andrijapanicsb commented:
2 x LGTM, tests passing fine.

Merging
